UCP: Register struct memory in UCT (draft)

brminich opened the pull request:
## What
Adding memory registration in UCT for UCP structs

brminich committed:
UCP: Register struct memory in UCT (draft)

artpol84 commented on src/ucp/core/ucp_request.c:
```diff
@@ -255,6 +259,39 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_request_memory_reg,
+        md_idx = ucp_ep_md_index(req_dbg->send.ep, lane);
+
+        /* register contig memory block covering the whole struct */
+        status = ucp_mem_rereg_mds(context, UCS_BIT(md_idx), buffer, 1024,//s->len*10,
```
`//s->len*10,`
Need to calculate the extent when the datatype is registered.
I'll take care of this.

artpol84 commented on src/ucp/dt/dt_struct.c:
```diff
@@ -17,34 +17,45 @@
             break;
         case UCP_DATATYPE_STRUCT:
             length += ucp_dt_struct_length(ucp_dt_struct(dsc->dt));
+            iovs   += ucp_dt_struct(dsc->dt)->uct_iov_count;
```
I think that you should add "1" here ...
Let's go over it when we will be sync'ing

artpol84 replied:
```C
if(repcount > 1 ){
   iov++;
} else {
   iovs   += ucp_dt_struct(dsc->dt)->uct_iov_count;
}
```

artpol84 commented on src/ucp/dt/dt_struct.c:
```diff
@@ -323,42 +328,91 @@ static int _is_primitive_strided(ucp_dt_struct_t *s)
+                iov->buffer = ptr;
+                iov->length = s_in->len;
+                iov->stride = s->desc[i].extent;
+                status      = _struct_register_ep_rec(ep, ptr, s_in,
```
For the sake of completeness, need to check 
```C
ucp_dt_struct_in_cache(s, buf);
```

artpol84 commented on src/ucp/proto/proto_am.inl:
```diff
@@ -214,6 +214,17 @@ void ucp_dt_iov_copy_uct(ucp_context_h context, uct_iov_t *iov, size_t *iovcnt,
                                             src_iov, length_max, md_index,
                                             md_flags);
         break;
+    case UCP_DATATYPE_STRUCT:
```
when is this function invoked?

artpol84 commented on src/uct/ib/mlx5/exp/ib_exp_md.c:
```diff
@@ -351,9 +377,9 @@ uct_ib_mlx5_exp_umr_alloc(uct_ib_mlx5_md_t *md, const uct_iov_t *iov,
-    } else { /* stride/interleave */
         status = uct_ib_mlx5_exp_umr_fill_region(umr, iov, iov_count);
+    } else { /* stride/interleave */
+        status = uct_ib_mlx5_exp_umr_fill_repeated(umr, iov, iov_count);
```
+1

artpol84 commented on src/ucp/dt/dt_struct.c:
```diff
@@ -323,42 +328,91 @@ static int _is_primitive_strided(ucp_dt_struct_t *s)
+            if (s_in->rep_count == 1) {
+                iov = _fill_uct_iov_rec(ep, ptr, s_in, contig_memh, iov);
+            } else {
+                iov->buffer = ptr;
```
On one hand, is more complicated in general case:
 If you have rep count > 1, you need to consider that a sub-type has to be repeated multiple times. So when you registering it, you need to register `s->rep_count` repetitions of this datatype and subsequently register with interleaved UMR where stride == sizeof(subdt).
On the other hand, we are not expecting to support this in our initial implementation. So, maybe just assert/abort this case.

artpol84 commented on src/ucp/dt/dt_struct.c:
```diff
@@ -323,42 +328,91 @@ static int _is_primitive_strided(ucp_dt_struct_t *s)
+        if (UCP_DT_IS_STRUCT(s->desc[i].dt)) {
+            s_in = ucp_dt_struct(s->desc[i].dt);
+            if (s_in->rep_count == 1) {
+                iov = _fill_uct_iov_rec(ep, ptr, s_in, contig_memh, iov);
```
Here it is possible that a sub-datatype is actually UMR.
This is what we see in LFRic: UMR/KLM that contains contig and UMR/repeated items.

artpol84 commented on src/ucp/dt/dt_struct.h:
```diff
@@ -65,18 +67,30 @@ static inline size_t ucp_dt_struct_depth(const ucp_dt_struct_t *s)
     return s->depth;
 }
 
+static UCS_F_ALWAYS_INLINE uct_mem_h ucp_dt_struct_in_cache(ucp_dt_struct_t *s,
```
Minor Again, IMO the cleaner solution is to hide CACHE inside struct.

artpol84 reviewed (commented):
Let's go over it together.

brminich committed:
UCP/UCT: Add noncontig registration on MD

brminich committed:
DT struct cleanup and gtest

- hash cleanup during dt struct destruction
- simple tests for plain and interleave types
